File: holoken-app/src/main/java/com/holokenmod/ui/MainDialogs.java
package com.holokenmod.ui;

import android.app.ActivityOptions;
import android.content.Intent;
import android.view.LayoutInflater;
import android.view.View;

import androidx.appcompat.app.AlertDialog;

import com.holokenmod.Game;
import com.holokenmod.R;

class MainDialogs {
	private final MainActivity mainActivity;
	private final Game game;
	
	MainDialogs(MainActivity mainActivity, Game game) {
		this.mainActivity = mainActivity;
		this.game = game;
	}
	
	void newGameGridDialog() {
		Intent intent = new Intent(mainActivity, NewGameActivity.class);
		
		if (game.getGridUI() != null && game.getGrid() != null) {
			ActivityOptions options = ActivityOptions.makeSceneTransitionAnimation(
					mainActivity,
					game.getGridUI(),
					"grid");
			
			mainActivity.startActivityForResult(intent, 0, options.toBundle());
		} else {
			mainActivity.startActivityForResult(intent, 0);
		}
	}
	
	void restartGameDialog() {
		if (!game.getGrid().isActive()) {
			return;
		}
		final AlertDialog.Builder builder = new AlertDialog.Builder(mainActivity);
		builder.setTitle(R.string.dialog_restart_title)
				.setMessage(R.string.dialog_restart_msg)
				.setIcon(android.R.drawable.ic_dialog_alert)
				.setNegativeButton(R.string.dialog_cancel, (dialog, id) -> dialog.cancel())
				.setPositiveButton(R.string.dialog_ok, (dialog, id) -> {
					game.clearUserValues();
					game.getGrid().setActive(true);
					mainActivity.startFreshGrid(true);
				})
				.show();
	}
	
	void openHelpDialog() {
		final AlertDialog.Builder builder = new AlertDialog.Builder(mainActivity);
		final LayoutInflater inflater = LayoutInflater.from(mainActivity);
		final View layout = inflater.inflate(R.layout.dialog_help,
				mainActivity.findViewById(R.id.help_layout));
		builder.setTitle(R.string.help_section_title)
				.setView(layout)
				.setNeutralButton(R.string.about_section_title, (dialog, id) -> this
						.openAboutDialog())
				.setPositiveButton(R.string.dialog_ok, (dialog, id) -> dialog.cancel())
				.show();
	}
	
	void openAboutDialog() {
		final AlertDialog.Builder builder = new AlertDialog.Builder(mainActivity);
		final LayoutInflater inflater = LayoutInflater.from(mainActivity);
		final View layout = inflater.inflate(R.layout.dialog_about,
				mainActivity.findViewById(R.id.about_layout));
		
		builder.setTitle(R.string.about_section_title)
				.setView(layout)
				.setNeutralButton(R.string.help_section_title, (dialog, id) -> this
						.openHelpDialog())
				.setPositiveButton(R.string.dialog_ok, (dialog, id) -> dialog.cancel())
				.show();
	}
}

File: holoken-app/src/main/kotlin/com/holokenmod/ui/MainDialogs.kt
package com.holokenmod.ui

import android.app.ActivityOptions
import android.content.DialogInterface
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import androidx.appcompat.app.AlertDialog
import com.holokenmod.R
import com.holokenmod.game.Game
import com.holokenmod.ui.main.MainActivity
import com.holokenmod.ui.newgame.NewGameActivity
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject

class MainDialogs(private val mainActivity: MainActivity): KoinComponent {
    private val game: Game by inject()

    fun newGameGridDialog() {
        val intent = Intent(mainActivity, NewGameActivity::class.java)
        val options = ActivityOptions.makeSceneTransitionAnimation(
            mainActivity,
            game.gridUI as View?,
            "grid"
        )
        mainActivity.startActivityForResult(intent, 0, options.toBundle())
    }

    fun restartGameDialog() {
        if (!game.grid.isActive) {
            return
        }
        val builder = AlertDialog.Builder(
            mainActivity
        )
        builder.setTitle(R.string.dialog_restart_title)
            .setMessage(R.string.dialog_restart_msg)
            .setIcon(android.R.drawable.ic_dialog_alert)
            .setNegativeButton(R.string.dialog_cancel) { dialog: DialogInterface, _: Int -> dialog.cancel() }
            .setPositiveButton(R.string.dialog_ok) { _: DialogInterface?, _: Int ->
                game.restartGame()
                mainActivity.startFreshGrid(true)
            }
            .show()
    }

    fun openHelpDialog() {
        val builder = AlertDialog.Builder(
            mainActivity
        )
        val inflater = LayoutInflater.from(mainActivity)
        val layout = inflater.inflate(
            R.layout.dialog_help,
            mainActivity.findViewById(R.id.help_layout)
        )
        builder.setTitle(R.string.help_section_title)
            .setView(layout)
            .setNeutralButton(R.string.about_section_title) { _: DialogInterface?, _: Int -> openAboutDialog() }
            .setPositiveButton(R.string.dialog_ok) { dialog: DialogInterface, _: Int -> dialog.cancel() }
            .show()
    }

    private fun openAboutDialog() {
        val builder = AlertDialog.Builder(
            mainActivity
        )
        val inflater = LayoutInflater.from(mainActivity)
        val layout = inflater.inflate(
            R.layout.dialog_about,
            mainActivity.findViewById(R.id.about_layout)
        )
        builder.setTitle(R.string.about_section_title)
            .setView(layout)
            .setNeutralButton(R.string.help_section_title) { _: DialogInterface?, _: Int -> openHelpDialog() }
            .setPositiveButton(R.string.dialog_ok) { dialog: DialogInterface, _: Int -> dialog.cancel() }
            .show()
    }
}